### PR TITLE
11 add docstrings

### PR DESCRIPTION
**Pull request by pie-tras**
Adds docstrings.

### CODE REVIEW

**Comment by sapols** [comment]
I ran this code through GPT-4 to get its opinions on the quality of the new docstrings and how well they follow PEP 257 standards. I'll comment the results here.

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:6]: "The module-level docstring at the start of your script is well written, but according to PEP 257, it should be a one-liner followed by a more detailed description if necessary. Something like:
```
"""
A script to retrieve and manipulate datasets from a Latis instance.

This script provides functionality to retrieve the dataset in either numpy 
or pandas format, and optionally download the dataset to a file.
"""
```
"

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:97]: "In the `download()` function, you are missing the argument `baseUrl` in the call to `__datasetWillUseVersion3()`. This isn't an issue with the docstring per se, but a potential bug in your code."

**Comment by sapols** [comment]
> The docstrings should describe what the function does in the present tense. So, "Check which version of Latis dataset will be used" should be "Checks which version of Latis dataset will be used.", "Retrieve dataset from Latis instance" should be "Retrieves dataset from Latis instance", and "Download dataset from Latis instance to file" should be "Downloads dataset from Latis instance to file".

I'll also add that we should be careful how we capitalize "LaTiS". I figure fully-lowercase `latis` is fine when referring to variables and function names, but otherwise we should use `LaTiS`. "Latis" is incorrect.

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:62]: > You should specify what type of elements are expected in the projections, selections, and operations lists. For example, if they are lists of strings:
> ```
> projections (List[str], optional): List of projections to use.
> selections (List[str], optional): List of selections to use.
> operations (List[str], optional): List of operations to apply.
> ```

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:93]: > You should specify what type of elements are expected in the projections, selections, and operations lists. For example, if they are lists of strings:
> ```
> projections (List[str], optional): List of projections to use.
> selections (List[str], optional): List of selections to use.
> operations (List[str], optional): List of operations to apply.
> ```

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:130]: > In the getDataset method, clarify the types of elements for projections, selections, and operations. As we discussed earlier, they are lists of strings, so indicate that in the docstring.

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:156]: > In the class-level docstring, specify that `datasets` is a dictionary mapping formatted dataset names to their Latis names:
> `datasets (dict of {str: str}): A dictionary mapping formatted dataset names to their Latis names.`

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:160]: It suggests: `latisInstance (LatisInstance): The Latis instance this catalog is associated with.`

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:191]: > In the `search` method: The searchTerm argument is not optional in the sense that the method expects it, but the behavior when searchTerm is an empty string or None might be non-obvious. Clarify this in the docstring:
>         ```

>         """Filters the catalog by a search term.
> 
>         If the search term is an empty string or None, returns all datasets in the catalog.
> 
>         Args:
>             searchTerm (str): The term to filter the catalog by.
> 
>         Returns:
>             list: A filtered list of dataset Latis names.
>         """
> ```

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:206]: > Like previous cases, clarify the types of elements for the projections, selections, and operations attributes in the class-level docstring. They should be lists of strings.

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:239]: > In the `select` method, the `inclusive` argument description can be a bit misleading as it reads as though both bounds would be inclusive if it's True. However, only the start bound will be inclusive in this case. It would be best to clarify that.

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:263]: It suggests: `projectionList (List[str]): List of projections.`

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:198]: Reminder to use present tense for the docstrings everywhere in this file.

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:332]: > In the `getFile` method, describe that it writes the dataset data to a local file. For example, you might add "Writes the dataset data to a local file."

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:376]: It suggests:
```
latisInstance (LatisInstance): The Latis instance from which to retrieve metadata.
dataset (Dataset): The dataset for which to retrieve metadata.
```

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:379]: In the `__init__` method, `latisInstance` and `dataset` could be explained in more detail. These arguments are critical to understanding how to use the class properly. Here's how I would revise the docstrings:

```
def __init__(self, latisInstance, dataset):
        """Initializes the Metadata object and retrieves the metadata properties.

        Retrieves metadata from the provided Latis instance and dataset. The
        retrieval method depends on the version of Latis being used.

        Args:
            latisInstance (LatisInstance): The Latis instance from which to
            retrieve metadata.
            dataset (Dataset): The dataset for which to retrieve metadata.
        """
```

**Review by sapols — CHANGES_REQUESTED** [review]
Overall the docstrings are great. This will be ready to merge after the suggested edits.

ChatGPT recently added a feature to share conversations, so if anyone is curious to see how I prompted it, here ya go: https://chat.openai.com/share/9931dfe2-62e9-4dea-9719-7bd9e77ee496

File: src/latis-python-client/latis.py
Comment (sapols) [reply src/latis-python-client/latis.py:269]: Now the docstring is correct, but I don't think this is the right implementation. As is, the user can't perform a select where the start is inclusive but the end is exclusive, which is the most common use case (and should arguably be the default behavior). The right move might be to have two `inclusive` vars for start and end respectively, although I'm not sure.

Thoughts, @dlindhol @lindholc ?

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:64]: Aren't projections just lists of strings? Why'd you label them as lists of lists?

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:412]: Didn't capitalize the 'S' in three 'LaTis's here.

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:55]: Should be "Retrieves"

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:88]: Should be "Downloads (a?) dataset"

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:262]: Should be "Defines"

File: src/latis-python-client/latis.py
Comment (sapols) [inline src/latis-python-client/latis.py:290]: Should be "Appends"

**Review by sapols — CHANGES_REQUESTED** [review]
Left a few more comments for discussion

File: src/latis-python-client/latis.py
Comment (dlindhol) [reply src/latis-python-client/latis.py:269]: I think that the start should always be inclusive. It's the end that is typically the problem. I would default to exclusive (to avoid the double hit when stringing together multiple contiguous requests).

Comment (lindholc) [reply src/latis-python-client/latis.py:269]: See also #18.

Comment (dlindhol) [reply src/latis-python-client/latis.py:269]: There are indeed cases where one might want the start to be exclusive. I suppose we shouldn't limit LaTiS capabilities.

File: src/latis-python-client/latis.py
Comment (pie-tras) [reply src/latis-python-client/latis.py:64]: Yep. Forgot `project()` appends each element not just the list.

**Comment by sapols** [comment]
Cool, LGTM minus the `select` inclusivity thing (which is a simple enough change that I think we should fix it before merging this).

File: src/latis-python-client/latis.py
Comment (pie-tras) [reply src/latis-python-client/latis.py:269]: Since issue #18 exists what should be implemented in this PR? I can leave it how it is for now or maybe just add a second bool to independently toggle the start and end bounds. @sapols

**Review by sapols — APPROVED** [review]
I like it. 👍

**Pull request merged**
